controller/api/destination/server.go

```diff
@@ -344,58 +324,132 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
+	if err != nil {
+		s.log.Errorf("failed to get opaque ports annotation for pod: %s", err)
+	}
 
-		skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(pod)
-		if err != nil {
-			log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+	// If the opaque ports annotation was not set, then use the destination
+	// server's default value.
+	if !ok {
+		opaquePorts = s.defaultOpaquePorts
+	}
+
+	skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(pod)
+	if err != nil {
+		s.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+	}
+	podSet := podToAddressSet(s.k8sAPI, pod).WithPort(port)
+	podID := watcher.PodID{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	}
+	endpoint, err := toWeightedAddr(podSet.Addresses[podID], opaquePorts, skippedInboundPorts, s.enableH2Upgrade, s.identityTrustDomain, s.controllerNS, s.log)
+	if err != nil {
+		return err
+	}
+
+	// `Get` doesn't include the namespace in the per-endpoint metadata, so it
+	// needs to be special-cased.
+	endpoint.MetricLabels["namespace"] = pod.Namespace
+
+	translator := newProfileTranslator(stream, s.log, "", port, endpoint)
+
+	// Create a client which watches for changes to the inbound policy for
+	// portSpec.
+	portSpec := &policyPb.PortSpec{
+		Workload: fmt.Sprintf("%s:%s", pod.Namespace, pod.Name),
+		Port:     port,
+	}
+	client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
+	if err != nil {
+		return err
+	}
+	s.log.Debugf("Establishing watch on policy server %s:%d", portSpec.Workload, portSpec.Port)
+
+	// Receive policy server updates for portSpec.
+	//
+	// Running this in the background allows it to handle Recv errors when
+	// the stream context is canceled. When canceled, this routine will
+	// complete.
+	updates := make(chan *policyPb.Server)
+	go func() {
+		for {
+			update, err := client.Recv()
+			if err != nil {
+				errStatus, _ := status.FromError(err)
+				if errStatus.Code() == codes.NotFound {
+					// If the policy server has not indexed the pod, we don't
+					// want to close the stream, but we also don't want to
+					// keep calling Recv since it will keep sending nil
+					// updates. Instead, send one update and stop.
+					updates <- nil
+					break
+				} else {
+					s.log.Debugf("Shutting down policy server updates for %s:%d: %s", portSpec.Workload, portSpec.Port, err)
+					close(updates)
+					break
+				}
+			}
+			updates <- update
 		}
+	}()
+
+	// For each policy server update, create a new destination profile for the
+	// endpoint. If the server update does not indicate that the protocol is
+	// opaque, fall back to the opaque ports annotation on the pod.
+	//
+	// Shutdown is handled by destination server shutdown or the stream being
+	// closed.
+	for {
+		select {
+		case <-s.shutdown:
+			return nil
+		case <-stream.Context().Done():
+			// This case handles shutdown when updates will never be closed.
```

Comment:
I don't really understand this comment.  It's hard to follow the scenario that it describes.  If the GetProfile request is cancelled, the policy request will also get cancelled because they share a context.  This should cause Recv to return an error and cause the updates stream to get closed.  We handle the channel being closed in the next case, so I don't think we actually need this case at all.

Reply:
The explicit handling of stream closure here is needed because of the policy server `NotFound` check [here](https://github.com/linkerd/linkerd2/blob/52f791ad252cc843759afb68811e5cce14ad17ba/controller/api/destination/server.go#L386-L393).

In this case, the policy server updates goroutine has completed without closing `updates`. This means that this goroutine would never complete _if_ we are only checking for `updates` to be closed. So to prevent this, we have to explicitly handle `stream` closing which allow this goroutine to complete.